sql: break dependencies on sql/parser

yuzefovich opened the pull request:
This PR contains multiple commits that break dependency on `sql/parser` from multiple packages that are on the critical path for builds. Optimizing `sql/parser` build time doesn't seem trivial, but doing more work in parallel with that is the next best option. The general idea is to inject the necessary parser functions into packages that need them at `init` time. See each commit for more details.

Informs: #79357.
Epic: None
Release note: None

yuzefovich committed:
sqlstats/insights: extract proto into a separate package

This allows us to break dependency of `serverpb` on `sql/parser` (which
was recently introduced when we added support for SQL Commenter query
tags).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/152478)
<!-- Reviewable:end -->

yuzefovich committed:
sem/eval: break dependency on sql/parser

This is achieved by injecting `parser.ParseOne` when initializing the
sql package.

Release note: None

yuzefovich committed:
sem/builtins: break dependency on sql/parser

This is achieved by injecting a couple of functions when initializing
the sql package.

Release note: None

yuzefovich committed:
catalog/funcdesc: break dependency on sql/parser

Release note: None

yuzefovich committed:
catalog/schemaexpr: break dependency on sql/parser

Release note: None

yuzefovich committed:
catalog/tabledesc: break dependency on sql/parser

Release note: None

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice. Do you think it would be more intuitive for code-reading to put the injected functions all in a single place, like `parserutils` maybe? Or would that defeat the purpose?

> @DrewKimball reviewed 20 of 20 files at r1, 3 of 3 files at r2, 11 of 11 files at r3, 4 of 4 files at r4, 22 of 22 files at r5, 10 of 10 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152478#-:-OYhfOIFDSyJcUoNnRB7:b-uco8c4)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Yeah, we cannot really do the injection from `parserutils` package - we need to do the injection in the "higher" level package (like `sql` that will always depend on `sql/parser` and that also depends on packages in which we do the injection) but `parserutils` package is a low level one, which shouldn't depend on `sql/parser` nor on others.

One thing I debated on whether it was worth extracting `DoParserInjection` function and then calling it from tests vs doing just the necessary injections from tests. The current way makes building _test_ packages that call `sql.DoParserInjection()` depend on `sql` too, slowing down the test builds. But given that only a handful of test packages are affected, I decided to keep it this way. If we start doing this "break dependency via `init` injection" in more places, and more test packages will need this, we might switch to a tigher test-injection logic so that test packages don't depend on `pkg/sql`.

TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152478#-:-OYhgW036b8c4nPOloR3:b-s6uzqk)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Oh, I guess I misunderstood your suggestion. Let me think about it a bit more.

bors r-

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152478#-:-OYhhnW4-j9bBcdhEciI:b2taxn7)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
Nice suggestion, I like it, thanks!

bors r+
